eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs

```diff
@@ -1,132 +1,424 @@
-  PrevGovActionIdsChildren (..),
-  pgacPParamUpdateL,
-  pgacHardForkL,
-  pgacCommitteeL,
-  pgacConstitutionL,
-  proposalsRemoveDescendentIds,
+  prevGovActionIdsL,
+  fromPrevGovActionIds,
+  toPrevGovActionIds,
   -- Testing
   isConsistent_,
 ) where
 
-import Cardano.Ledger.Babbage.Core
-import Cardano.Ledger.BaseTypes (StrictMaybe (..))
-import Cardano.Ledger.Binary (DecCBOR (..), DecShareCBOR (..), EncCBOR (..))
-import Cardano.Ledger.Binary.Coders (
-  Decode (..),
-  Encode (..),
-  decode,
-  encode,
-  (!>),
-  (<!),
+import Cardano.Ledger.BaseTypes (
+  StrictMaybe (..),
+  maybeToStrictMaybe,
+  strictMaybeToMaybe,
  )
-import Cardano.Ledger.Conway.Governance.Procedures (
-  GovAction (..),
-  GovActionId,
-  GovActionPurpose (..),
-  GovActionState (..),
-  PrevGovActionId (PrevGovActionId),
-  Vote,
-  Voter (..),
-  gasActionL,
-  gasChildrenL,
-  gasCommitteeVotesL,
-  gasDRepVotesL,
-  gasIdL,
-  gasStakePoolVotesL,
+import Cardano.Ledger.Binary (
+  DecCBOR (..),
+  DecShareCBOR (..),
+  EncCBOR (..),
+  decodeRecordNamed,
+  encodeListLen,
  )
+import Cardano.Ledger.Binary.Coders (Decode (From, RecD), Encode (Rec, To), decode, encode, (!>), (<!))
+import Cardano.Ledger.Conway.Governance.Procedures
+import Cardano.Ledger.Core
 import Control.DeepSeq (NFData)
+import Control.Exception (assert)
 import Control.Monad (guard)
-import Data.Aeson (KeyValue, ToJSON (..), object, pairs, (.=))
+import Data.Aeson (KeyValue ((.=)), ToJSON (..), object, pairs)
 import Data.Default.Class (Default (..))
+import Data.Foldable (foldl')
 import Data.Functor (($>))
+import Data.Kind (Type)
 import qualified Data.Map as Map
 import Data.Map.Strict (Map)
+import Data.Maybe (isNothing)
 import qualified Data.OMap.Strict as OMap
+import Data.Sequence (Seq)
 import Data.Sequence.Strict (StrictSeq (..))
 import Data.Set (Set)
 import qualified Data.Set as Set
 import GHC.Generics (Generic)
 import Lens.Micro
 import NoThunks.Class (NoThunks)
 
-newtype Proposals era
-  = Proposals
-      (OMap.OMap (GovActionId (EraCrypto era)) (GovActionState era))
-  deriving newtype (Show, Eq)
-  deriving stock (Generic)
+data PRoot a = PRoot
+  { prRoot :: !(Maybe a)
+  , prChildren :: !(Set a)
+  }
+  deriving (Show, Eq, Ord, Generic, NoThunks, NFData, Default)
+
+prRootL :: Lens' (PRoot a) (Maybe a)
+prRootL = lens prRoot $ \x y -> x {prRoot = y}
+
+prChildrenL :: Lens' (PRoot a) (Set a)
+prChildrenL = lens prChildren $ \x y -> x {prChildren = y}
+
+data PNode a = PNode
+  { pnParent :: !(Maybe a)
+  , pnChildren :: !(Set a)
+  }
+  deriving (Show, Eq, Ord, Generic, NoThunks, NFData, Default)
+
+pnChildrenL :: Lens' (PNode a) (Set a)
+pnChildrenL = lens pnChildren $ \x y -> x {pnChildren = y}
+
+newtype PHierarchy a = PHierarchy
+  { unPHierarchy :: Map a (PNode a)
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+  deriving newtype (NoThunks, NFData, Default)
+
+pHierarchyNTL :: Lens' (PHierarchy a) (Map a (PNode a)) -- FIXME: what a terrible name!
```

Comment:
`pHierarchyNodesL` could work I think, a little less `wtf` than `NTL` perhaps :grinning:

Reply:
Agree :100: , I will get to it in the next push :+1:

Reply:
This is done :+1: Super thanks!